internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -46,67 +55,201 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
 		return nil, errors.New(err, "failed to unpack configuration")
 	}
 	return &contextProviderK8sSecrets{
-		logger: logger,
-		config: &cfg,
+		logger:       logger,
+		config:       &cfg,
+		secretsCache: make(map[string]*secretsData),
 	}, nil
 }
 
 func (p *contextProviderK8sSecrets) Fetch(key string) (string, bool) {
-	// key = "kubernetes_secrets.somenamespace.somesecret.value"
+	if p.config.DisableCache {
+		valid := p.validateKey(key)
+		if valid {
+			return p.fetchSecretWithTimeout(key)
+		} else {
+			return "", false
+		}
+	} else {
+		return p.getFromCache(key)
+	}
+}
+
+// Run initializes the k8s secrets context provider.
+func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
+	if err != nil {
+		p.logger.Debugf("Kubernetes_secrets provider skipped, unable to connect: %s", err)
+		return nil
+	}
 	p.clientMx.Lock()
-	client := p.client
+	p.client = client
 	p.clientMx.Unlock()
-	if client == nil {
-		return "", false
+
+	if !p.config.DisableCache {
+		go p.updateSecrets(ctx)
 	}
+
+	<-comm.Done()
+
+	p.clientMx.Lock()
+	p.client = nil
+	p.clientMx.Unlock()
+	return comm.Err()
+}
+
+func getK8sClient(kubeconfig string, opt kubernetes.KubeClientOptions) (k8sclient.Interface, error) {
+	return kubernetes.GetKubernetesClient(kubeconfig, opt)
+}
+
+// Update the secrets in the cache every RefreshInterval
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	timer := time.NewTimer(p.config.RefreshInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			p.updateCache()
+			timer.Reset(p.config.RefreshInterval)
+		}
+	}
+}
+
+func (p *contextProviderK8sSecrets) updateCache() {
+	// deleting entries does not free the memory, so we need to create a new map
+	// to place the secrets we want to keep
+	cacheTmp := make(map[string]*secretsData)
+
+	p.secretsCacheMx.Lock()
+	for name, data := range p.secretsCache {
+		diff := time.Since(data.lastAccess)
+		if diff < p.config.TTLDelete {
+			value, ok := p.fetchSecretWithTimeout(name)
+			if ok {
+				newData := &secretsData{
+					value:      value,
+					lastAccess: data.lastAccess,
+				}
+				cacheTmp[name] = newData
+			}
+
+		}
+	}
+	p.secretsCache = cacheTmp
+	p.secretsCacheMx.Unlock()
```

Comment:
@blakerouse I can understand your concern. For now the option to disable the cache has been added [0f8d86a](https://github.com/elastic/elastic-agent/pull/3822/commits/0f8d86a4955f6737e74b504e83f11a6188fcb3fb)

So in case of any problem like you describe we can revert to ption where we just get the secrret [here](https://github.com/elastic/elastic-agent/blob/0f8d86a4955f6737e74b504e83f11a6188fcb3fb/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L240) with a small lock to get the [client](https://github.com/elastic/elastic-agent/blob/0f8d86a4955f6737e74b504e83f11a6188fcb3fb/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L223)

Can we agree that Syncronisation of cache can be  done in a new story? We also investigate this on other cases for kubernetes watchers and if we have any update we can link this issue as well. 
For now I will approve this story

Reply:
I agree that holding the lock during the whole duration of update is not a good solution due to the blocking it can create to Fetch. I understand from the code that the reason to lock, is because we read all the keys and the lastaccess from the cache in the iteration. We actually write to the cache only after all secrets are updated. Why don't we keep the keys of the cache map in a slice and only lock the cache when we actually read the last access value for each key. Something like this 

```
func (p *contextProviderK8sSecrets) updateCache() {
	cacheTmp := make(map[string]*secretsData)

	p.secretsCacheMx.Lock()
	secretKeys := make([]string, len(p.secretsCache))
	i := 0
	for k := range p.secretsCache {
		secretKeys[i] = k
		i++
	}
	p.secretsCacheMx.Unlock()
	for _, name := range secretKeys {
		p.secretsCacheMx.Lock()
		lastAccess := p.secretsCache[name].lastAccess
		p.secretsCacheMx.Unlock()
		diff := time.Since(lastAccess)

		if diff < p.config.TTLDelete {
			value, ok := p.fetchSecretWithTimeout(name)
			if ok {
				newData := &secretsData{
					value:      value,
					lastAccess: lastAccess,
				}
				cacheTmp[name] = newData
			}

		}
	}
	p.secretsCacheMx.Lock()
	p.secretsCache = cacheTmp
	p.secretsCacheMx.Unlock()
}
```

Reply:
That has already been suggested, and @cmacknz explains in this https://github.com/elastic/elastic-agent/pull/3822#discussion_r1416314549 why that does not work. @MichaelKatsoulis

Reply:
You mean that they keys can change by fetch method which can create a new entry in the cache, and then in update due to the temp cache we will override and lose the new key? 
Because if this is the problem, it could be mitigated if we then merge the `p.secretsCache` and `cacheTmp` while holding the cache.

Reply:
> You mean that they keys can change by fetch method which can create a new entry in the cache, and then in update due to the temp cache we will override and lose the new key?

Yes.

> Because if this is the problem, it could be mitigated if we then merge the p.secretsCache and cacheTmp while holding the cache.

We cannot do that. It is explained in this https://github.com/elastic/elastic-agent/pull/3822#discussion_r1406557666 why not

Reply:
In that comment it is explained that deleting a key from a map does not delete it from memory. I am saying that after the update is done and new keys are added in the temp cache we can do a merge of the temp cache and p.secretsCache and keep the keys that may have been added while we were updating. See this example:

```
p.secretsCacheMx.Lock()
newCache := mergeMaps(p.secretsCache, cacheTmp, p.config.TTLDelete)
p.secretsCache = newCache
p.secretsCacheMx.Unlock()
```

```
func mergeMaps[K comparable, V any](m1 map[K]V, m2 map[K]V, ttl time.Duration) map[K]V {
	merged := make(map[K]V)
	for key, value := range m1 {
		lastAccess := any(value).(secretsData).lastAccess
		diff := time.Since(lastAccess)
		if diff < ttl {
			merged[key] = value
		}
	}
	for key, value := range m2 {
		merged[key] = value
	}
	return merged
}
```

Reply:
I agree with @MichaelKatsoulis suggestion but I believe you need to make an approvement to the mergeMaps. I believe that it is updating the lastAccess time from the `m2` map. It should most likely only overwrite the value and not the `lastAccess` timestamp.

Reply:
Yeah, in case there was a key accessed by fetch, while the update took place, then the last access of that key will have been updated. And by 
```
for key, value := range m2 {
		merged[key] = value
	}
```
the new last access will be overridden by lets say the old last access of the cache. 
Maybe using generics in this use case does not add anything and makes it more complicated as we already know the data types of the maps.
Something like this could work
```
func mergeMaps(m1, m2 map[string]*secretsData, ttl time.Duration) map[string]*secretsData {
	merged := make(map[string]*secretsData)
	for key, value := range m1 {
		lastAccess := value.lastAccess
		diff := time.Since(lastAccess)
		if diff < ttl {
			merged[key] = value
		}
	}
	for key, value := range m2 {
		if val, ok := merged[key]; ok {
			merged[key] = &secretsData{
				value:      value.value,
				lastAccess: val.lastAccess,
			}
		} else {
			merged[key] = value
		}
	}
	return merged
}
```

Reply:
I don't believe you want that last else statement. If the key was removed due to ttl in the previous loop you don't want to add it back in the case that it was fetched in the cache refresh, otherwise I believe the function will work and prevent the holding of the lock for the full refresh.

Reply:
I made a new commit, so now:

1. We hold only a read lock to update the secrets.
2. After that, we merge the maps.

Is it ok this way @blakerouse @MichaelKatsoulis ?

Reply:
@blakerouse you are right. I somehow thought that update adds extra keys as well but that is not the case.

Reply:
It is a reading lock now @blakerouse (this file preview is not updated to the last commit)

Reply:
I updated the code to do that in the latest commit @blakerouse